fastparquet/api.py

```diff
@@ -371,6 +376,15 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
             (and any contained files). Not required if this ParquetFile has
             a .fs file system attribute
         """
+        if not isinstance(rgs, list):
+            if not hasattr(rgs, '__iter__'):
+                # Filter tuple, generator.
+                rgs = [rgs]
+            else:
+                # If generator, switch to list as this list is re-used.
+                rgs = list(rgs)
+        if not rgs:
```

Comment:
Probably do this before the type checks, so that we don't end up trying to work with something like `[( )]` here.

Reply:
Martin, your 1st comment is on the outadated code:
_Does the first branch ever get called? A tuple also has __iter__, actually many things do. If what you really want to check for generator, why not do exactly that (inspect.isgenerator) ?_

Please, check you review the last version, which reads:
```python
        if not isinstance(rgs, list):
            if isinstance(rgs, ThriftObject):
                rgs = [rgs]
            else:
                # Use `list()` here, not `[]`, as the latter does not transform
                # generator into list but encapsulates generator in a list.
                rgs = list(rgs)
```
(I confirm it is correctly loaded in github, this copy/paste is directly from github, in tab `files` of this repo)

This should answer your comment, the 'troublesome' case if when provided `rgs`is a simple `ThriftObect`. The other cases are correctly handled by `list()`, but not the case when `rgs` is a `ThriftObject`.

Reply:
OK, got it. If `dict` is also acceptable for the condition (because a `ThriftObject` is just a wrapper for `dict`), please also allow that.

Reply:
> OK, got it. If `dict` is also acceptable for the condition (because a `ThriftObject` is just a wrapper for `dict`), please also allow that.

I am sorry I don't understand.
`rgs` can be:
 - a row group (in which case it is a `ThriftObject`)
 - or an iterable of row groups (generator, list, tuple...)

I cannot imagine a case when a user would provide a `dict`, except if fastparquet, elsewhere, can also accept indifferently `dict`  as `ThriftObject`. Is it the case? If yes, then yes, let's also correctly manage this type with `[]`and not `list()` (on a `dict`, `list()` would only keep the keys of the dict, not encapsulate the dict in a list indeed)

Reply:
> As you have noticed, fmd.row_groups returns a list of ThriftObjects with row-group type, but fmd[3] returns a list of dicts. Both have the same information. The mapping between field names and integer index is in the specs constant in cencoding.pyx. I don't know if we ever expect the user to manipulate the raw dicts, but the code in fastparquet does, i.e., those portions where I commented in this PR to keep the index notation rather than attribute access, because it is (much) faster.

Ok, clear, so I have added `idct` type checking as well.